codex/blockexchange/engine/engine.nim

```diff
@@ -100,124 +93,147 @@ type
     price*: UInt256
 
 # attach task scheduler to engine
-proc scheduleTask(b: BlockExcEngine, task: BlockExcPeerCtx): bool {.gcsafe.} =
-  b.taskQueue.pushOrUpdateNoWait(task).isOk()
+proc scheduleTask(self: BlockExcEngine, task: BlockExcPeerCtx): bool {.gcsafe.} =
+  self.taskQueue.pushOrUpdateNoWait(task).isOk()
 
-proc blockexcTaskRunner(b: BlockExcEngine) {.async: (raises: []).}
+proc blockexcTaskRunner(self: BlockExcEngine) {.async: (raises: []).}
 
-proc start*(b: BlockExcEngine) {.async.} =
+proc start*(self: BlockExcEngine) {.async.} =
   ## Start the blockexc task
   ##
 
-  await b.discovery.start()
-  await b.advertiser.start()
+  await self.discovery.start()
+  await self.advertiser.start()
 
-  trace "Blockexc starting with concurrent tasks", tasks = b.concurrentTasks
-  if b.blockexcRunning:
+  trace "Blockexc starting with concurrent tasks", tasks = self.concurrentTasks
+  if self.blockexcRunning:
     warn "Starting blockexc twice"
     return
 
-  b.blockexcRunning = true
-  for i in 0 ..< b.concurrentTasks:
-    let fut = b.blockexcTaskRunner()
-    b.trackedFutures.track(fut)
-    asyncSpawn fut
+  self.blockexcRunning = true
+  for i in 0 ..< self.concurrentTasks:
+    let fut = self.blockexcTaskRunner()
+    self.trackedFutures.track(fut)
 
-proc stop*(b: BlockExcEngine) {.async.} =
+proc stop*(self: BlockExcEngine) {.async.} =
   ## Stop the blockexc blockexc
   ##
 
-  await b.discovery.stop()
-  await b.advertiser.stop()
+  await self.trackedFutures.cancelTracked()
+  await self.network.stop()
+  await self.discovery.stop()
+  await self.advertiser.stop()
 
   trace "NetworkStore stop"
-  if not b.blockexcRunning:
+  if not self.blockexcRunning:
     warn "Stopping blockexc without starting it"
     return
 
-  b.blockexcRunning = false
-  await b.trackedFutures.cancelTracked()
+  self.blockexcRunning = false
 
   trace "NetworkStore stopped"
 
 proc sendWantHave(
-    b: BlockExcEngine, addresses: seq[BlockAddress], peers: seq[BlockExcPeerCtx]
+    self: BlockExcEngine, addresses: seq[BlockAddress], peers: seq[BlockExcPeerCtx]
 ): Future[void] {.async.} =
   for p in peers:
     let toAsk = addresses.filterIt(it notin p.peerHave)
```

Comment:
I know you did not touch that line, but it would be cleaner to have `sendWantHave` not to do any filtering (I mean, this should happen elsewhere, where it is really relevant). Especially that `sendWantBlock` is nice and clean without side-effected (the caller expects the request to be just sent)... Maybe a small separate PR...